File: examples/mcp_client.py
#!/usr/bin/env python3
"""
Real MCP Client for Crystal MCP Server

This client connects to the Crystal MCP Server via stdio using the actual
MCP (Model Context Protocol) JSON-RPC format.

This is NOT a shortcut - this sends real MCP protocol messages to the server.
"""

import subprocess
import json
import sys
import os
from pathlib import Path
from typing import Optional, Dict, Any
import threading
import queue

PROJECT_ROOT = Path(__file__).parent.parent


class MCPClient:
    """Real MCP Client that communicates with the server via stdio."""

    def __init__(self):
        self.process: Optional[subprocess.Popen] = None
        self.request_id = 0
        self.response_queue = queue.Queue()
        self.reader_thread: Optional[threading.Thread] = None
        self.running = False

    def start_server(self):
        """Start the MCP server as a subprocess."""
        server_path = PROJECT_ROOT / "dist" / "index.js"

        if not server_path.exists():
            raise FileNotFoundError(f"Server not built. Run 'npm run build' first. Missing: {server_path}")

        env = os.environ.copy()
        env["PYTHON_PATH"] = sys.executable

        self.process = subprocess.Popen(
            ["node", str(server_path)],
            stdin=subprocess.PIPE,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            cwd=str(PROJECT_ROOT),
            env=env,
            text=True,
            bufsize=1
        )

        self.running = True

        # Start reader thread for stderr (server logs)
        self.stderr_thread = threading.Thread(target=self._read_stderr, daemon=True)
        self.stderr_thread.start()

        # Start reader thread for stdout (MCP responses)
        self.reader_thread = threading.Thread(target=self._read_responses, daemon=True)
        self.reader_thread.start()

        print("✓ MCP Server started")

    def _read_stderr(self):
        """Read server logs from stderr."""
        while self.running and self.process:
            line = self.process.stderr.readline()
            if line:
                print(f"[SERVER] {line.rstrip()}")
            elif self.process.poll() is not None:
                break

    def _read_responses(self):
        """Read MCP responses from stdout."""
        while self.running and self.process:
            line = self.process.stdout.readline()
            if line:
                try:
                    response = json.loads(line.strip())
                    self.response_queue.put(response)
                except json.JSONDecodeError:
                    print(f"[RAW] {line.rstrip()}")
            elif self.process.poll() is not None:
                break

    def _send_request(self, method: str, params: Optional[Dict] = None) -> Dict[str, Any]:
        """Send an MCP JSON-RPC request and wait for response."""
        self.request_id += 1

        request = {
            "jsonrpc": "2.0",
            "id": self.request_id,
            "method": method,
        }
        if params:
            request["params"] = params

        request_json = json.dumps(request)
        print(f"\n>>> SENDING MCP REQUEST:")
        print(f"    {request_json[:200]}{'...' if len(request_json) > 200 else ''}")

        self.process.stdin.write(request_json + "\n")
        self.process.stdin.flush()

        # Wait for response
        try:
            response = self.response_queue.get(timeout=60)
            print(f"\n<<< RECEIVED MCP RESPONSE:")
            response_str = json.dumps(response)
            print(f"    {response_str[:500]}{'...' if len(response_str) > 500 else ''}")
            return response
        except queue.Empty:
            return {"error": "Timeout waiting for response"}

    def initialize(self):
        """Send MCP initialize request."""
        print("\n" + "="*60)
        print("STEP 1: MCP INITIALIZE")
        print("="*60)

        response = self._send_request("initialize", {
            "protocolVersion": "2024-11-05",
            "capabilities": {
                "roots": {"listChanged": True}
            },
            "clientInfo": {
                "name": "crystal-mcp-test-client",
                "version": "1.0.0"
            }
        })

        if "result" in response:
            print("\n✓ Server initialized successfully")
            print(f"  Server: {response['result'].get('serverInfo', {}).get('name', 'Unknown')}")
            print(f"  Version: {response['result'].get('serverInfo', {}).get('version', 'Unknown')}")

            # Send initialized notification
            self._send_notification("notifications/initialized", {})

        return response

    def _send_notification(self, method: str, params: Dict = None):
        """Send an MCP notification (no response expected)."""
        notification = {
            "jsonrpc": "2.0",
            "method": method,
        }
        if params:
            notification["params"] = params

        self.process.stdin.write(json.dumps(notification) + "\n")
        self.process.stdin.flush()

    def list_tools(self):
        """List available MCP tools."""
        print("\n" + "="*60)
        print("STEP 2: LIST AVAILABLE TOOLS")
        print("="*60)

        response = self._send_request("tools/list", {})

        if "result" in response and "tools" in response["result"]:
            tools = response["result"]["tools"]
            print(f"\n✓ Found {len(tools)} MCP tools:")
            for tool in tools[:10]:  # Show first 10
                print(f"  • {tool['name']}: {tool.get('description', '')[:60]}...")
            if len(tools) > 10:
                print(f"  ... and {len(tools) - 10} more")

        return response

    def call_tool(self, tool_name: str, arguments: Dict[str, Any]):
        """Call an MCP tool."""
        print("\n" + "="*60)
        print(f"STEP 3: CALL TOOL - {tool_name}")
        print("="*60)
        print(f"Arguments: {json.dumps(arguments, indent=2)}")

        response = self._send_request("tools/call", {
            "name": tool_name,
            "arguments": arguments
        })

        return response

    def stop(self):
        """Stop the MCP server."""
        self.running = False
        if self.process:
            self.process.terminate()
            self.process.wait()
            print("\n✓ MCP Server stopped")


def save_output_files(result: Dict, output_dir: Path, name: str):
    """Save the generated structure to files."""
    output_dir.mkdir(exist_ok=True)

    if "result" in result and "content" in result["result"]:
        content = result["result"]["content"]
        for item in content:
            if item.get("type") == "text":
                text = item.get("text", "")
                # Parse JSON from <json-data> tags (new format)
                if "<json-data>" in text:
                    start = text.find("<json-data>") + 11
                    end = text.find("</json-data>")
                    try:
                        data = json.loads(text[start:end].strip())
                        if data.get("success") and "structure" in data:
                            # Save JSON
                            json_file = output_dir / f"{name}.json"
                            with open(json_file, 'w') as f:
                                json.dump(data, f, indent=2)
                            print(f"  ✓ Saved: {json_file}")

                            # Save CIF file if available
                            if data.get("files", {}).get("cif"):
                                cif_file = output_dir / f"{name}.cif"
                                with open(cif_file, 'w') as f:
                                    f.write(data["files"]["cif"])
                                print(f"  ✓ Saved: {cif_file}")

                            # Save XYZ file if available
                            if data.get("files", {}).get("xyz"):
                                xyz_file = output_dir / f"{name}.xyz"
                                with open(xyz_file, 'w') as f:
                                    f.write(data["files"]["xyz"])
                                print(f"  ✓ Saved: {xyz_file}")

                            return data
                    except json.JSONDecodeError as e:
                        print(f"  ✗ JSON parse error: {e}")
                # Fallback: try direct JSON parse
                else:
                    try:
                        data = json.loads(text)
                        if data.get("success") and "structure" in data:
                            json_file = output_dir / f"{name}.json"
                            with open(json_file, 'w') as f:
                                json.dump(data, f, indent=2)
                            print(f"  ✓ Saved: {json_file}")
                            return data
                    except json.JSONDecodeError:
                        pass
    return None


def main():
    """Run the real MCP client test."""
    print("="*70)
    print("   REAL MCP CLIENT TEST - Crystal Structure Generator")
    print("   Using actual MCP Protocol over stdio")
    print("="*70)

    client = MCPClient()
    output_dir = Path(__file__).parent / "mcp_real_outputs"

    try:
        # Start the MCP server
        client.start_server()

        # Wait for server to initialize
        import time
        time.sleep(2)

        # Step 1: Initialize
        init_response = client.initialize()
        if "error" in init_response:
            print(f"ERROR: {init_response['error']}")
            return

        time.sleep(0.5)

        # Step 2: List tools
        tools_response = client.list_tools()

        time.sleep(0.5)

        # Step 3: Generate a crystal structure using REAL MCP call
        print("\n" + "="*60)
        print("GENERATING CRYSTAL STRUCTURE VIA MCP")
        print("="*60)

        # Silicon Diamond structure
        print("\n>>> Natural Language Intent: 'Generate silicon in diamond structure'")
        print(">>> Translated to MCP tool call: generate_crystal")

        result = client.call_tool("generate_crystal", {
            "composition": ["Si", "Si", "Si", "Si", "Si", "Si", "Si", "Si"],
            "space_group": 227,
            "seed": 42
        })

        if result:
            data = save_output_files(result, output_dir, "silicon_diamond_mcp")
            if data and data.get("success"):
                structure = data["structure"]
                print(f"\n✓ Structure generated via MCP:")
                print(f"  Formula: {structure['metadata']['formula']}")
                print(f"  Space Group: {structure['space_group']['symbol']} (#{structure['space_group']['number']})")
                print(f"  Atoms: {structure['metadata']['natoms']}")
                print(f"  Volume: {structure['lattice']['volume']:.2f} Å³")

        # Step 4: Generate another structure
        print("\n>>> Natural Language Intent: 'Create NaCl rocksalt for ionic study'")
        print(">>> Translated to MCP tool call: generate_crystal")

        result2 = client.call_tool("generate_crystal", {
            "composition": ["Na", "Na", "Na", "Na", "Cl", "Cl", "Cl", "Cl"],
            "space_group": 225,
            "seed": 42
        })

        if result2:
            data2 = save_output_files(result2, output_dir, "nacl_rocksalt_mcp")
            if data2 and data2.get("success"):
                structure = data2["structure"]
                print(f"\n✓ Structure generated via MCP:")
                print(f"  Formula: {structure['metadata']['formula']}")
                print(f"  Space Group: {structure['space_group']['symbol']} (#{structure['space_group']['number']})")

        print("\n" + "="*60)
        print("MCP CLIENT TEST COMPLETE")
        print("="*60)
        print(f"\nOutput files saved to: {output_dir}")

    except FileNotFoundError as e:
        print(f"\nERROR: {e}")
        print("Please run 'npm run build' first to compile the TypeScript server.")
    except Exception as e:
        print(f"\nERROR: {e}")
        import traceback
        traceback.print_exc()
    finally:
        client.stop()


if __name__ == "__main__":
    main()

File: crystal-gui/src/mcp_client.rs
//! MCP Client for communicating with crystal-mcp-server
//! 
//! Implements JSON-RPC over stdio to communicate with the MCP server.

use std::io::{BufRead, BufReader, BufWriter, Write};
use std::process::{Child, ChildStdin, ChildStdout, Command, Stdio};
use std::sync::atomic::{AtomicU64, Ordering};

use serde::{Deserialize, Serialize};
use serde_json::{json, Value};
use thiserror::Error;

#[derive(Error, Debug)]
pub enum McpError {
    #[error("Failed to spawn MCP server: {0}")]
    SpawnError(String),
    #[error("Failed to send request: {0}")]
    SendError(String),
    #[error("Failed to read response: {0}")]
    ReadError(String),
    #[error("JSON parse error: {0}")]
    JsonError(#[from] serde_json::Error),
    #[error("MCP error: {code} - {message}")]
    McpError { code: i64, message: String },
    #[error("Server not initialized")]
    NotInitialized,
    #[error("IO error: {0}")]
    IoError(#[from] std::io::Error),
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Tool {
    pub name: String,
    #[serde(default)]
    pub description: Option<String>,
    #[serde(rename = "inputSchema", default)]
    pub input_schema: Option<Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ToolResult {
    pub content: Vec<ContentItem>,
    #[serde(rename = "isError", default)]
    pub is_error: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ContentItem {
    #[serde(rename = "type")]
    pub content_type: String,
    pub text: String,
}

pub struct McpClient {
    process: Option<Child>,
    stdin: Option<BufWriter<ChildStdin>>,
    stdout: Option<BufReader<ChildStdout>>,
    request_id: AtomicU64,
    initialized: bool,
    tools: Vec<Tool>,
    server_path: String,
}

impl McpClient {
    pub fn new(server_path: String) -> Self {
        Self {
            process: None,
            stdin: None,
            stdout: None,
            request_id: AtomicU64::new(1),
            initialized: false,
            tools: Vec::new(),
            server_path,
        }
    }

    fn next_id(&self) -> u64 {
        self.request_id.fetch_add(1, Ordering::SeqCst)
    }

    pub fn start(&mut self) -> Result<(), McpError> {
        let mut child = Command::new("node")
            .arg(&self.server_path)
            .stdin(Stdio::piped())
            .stdout(Stdio::piped())
            .stderr(Stdio::null())
            .spawn()
            .map_err(|e| McpError::SpawnError(format!("{}: {}", e, self.server_path)))?;

        let stdin = child.stdin.take().ok_or(McpError::SendError("Failed to get stdin".to_string()))?;
        let stdout = child.stdout.take().ok_or(McpError::ReadError("Failed to get stdout".to_string()))?;
        
        self.stdin = Some(BufWriter::new(stdin));
        self.stdout = Some(BufReader::new(stdout));
        self.process = Some(child);
        
        self.initialize()?;
        self.list_tools()?;
        Ok(())
    }

    fn send_request(&mut self, method: &str, params: Value) -> Result<Value, McpError> {
        let id = self.next_id();
        let request = json!({
            "jsonrpc": "2.0",
            "id": id,
            "method": method,
            "params": params
        });

        let stdin = self.stdin.as_mut().ok_or(McpError::NotInitialized)?;
        let request_str = serde_json::to_string(&request)?;
        writeln!(stdin, "{}", request_str).map_err(|e| McpError::SendError(e.to_string()))?;
        stdin.flush().map_err(|e| McpError::SendError(e.to_string()))?;

        let stdout = self.stdout.as_mut().ok_or(McpError::NotInitialized)?;
        let mut line = String::new();
        stdout.read_line(&mut line).map_err(|e| McpError::ReadError(e.to_string()))?;

        if line.is_empty() {
            return Err(McpError::ReadError("Empty response from server".to_string()));
        }

        let response: Value = serde_json::from_str(&line)?;
        
        if let Some(error) = response.get("error") {
            let code = error.get("code").and_then(|c| c.as_i64()).unwrap_or(-1);
            let message = error.get("message").and_then(|m| m.as_str()).unwrap_or("Unknown error").to_string();
            return Err(McpError::McpError { code, message });
        }

        Ok(response.get("result").cloned().unwrap_or(Value::Null))
    }

    fn send_notification(&mut self, method: &str) -> Result<(), McpError> {
        let stdin = self.stdin.as_mut().ok_or(McpError::NotInitialized)?;
        
        let notification = json!({
            "jsonrpc": "2.0",
            "method": method
        });

        let notification_str = serde_json::to_string(&notification)?;
        writeln!(stdin, "{}", notification_str).map_err(|e| McpError::SendError(e.to_string()))?;
        stdin.flush().map_err(|e| McpError::SendError(e.to_string()))?;
        Ok(())
    }

    fn initialize(&mut self) -> Result<(), McpError> {
        let params = json!({
            "protocolVersion": "2024-11-05",
            "capabilities": {},
            "clientInfo": {
                "name": "crystal-gui",
                "version": "0.1.0"
            }
        });

        let _result = self.send_request("initialize", params)?;
        self.send_notification("notifications/initialized")?;
        self.initialized = true;
        Ok(())
    }

    fn list_tools(&mut self) -> Result<(), McpError> {
        let result = self.send_request("tools/list", json!({}))?;
        
        // MCP returns {"tools": [...]}
        if let Some(tools_array) = result.get("tools").and_then(|t| t.as_array()) {
            self.tools = tools_array
                .iter()
                .filter_map(|t| {
                    // Extract just name and description
                    let name = t.get("name")?.as_str()?.to_string();
                    let description = t.get("description").and_then(|d| d.as_str()).map(|s| s.to_string());
                    let input_schema = t.get("inputSchema").cloned();
                    Some(Tool { name, description, input_schema })
                })
                .collect();
        }
        Ok(())
    }

    pub fn get_tools(&self) -> &[Tool] {
        &self.tools
    }

    pub fn call_tool(&mut self, name: &str, arguments: Value) -> Result<ToolResult, McpError> {
        let params = json!({
            "name": name,
            "arguments": arguments
        });

        let result = self.send_request("tools/call", params)?;
        let tool_result: ToolResult = serde_json::from_value(result)?;
        Ok(tool_result)
    }

    pub fn is_initialized(&self) -> bool {
        self.initialized
    }

    pub fn stop(&mut self) {
        if let Some(mut process) = self.process.take() {
            let _ = process.kill();
        }
        self.initialized = false;
    }
}

impl Drop for McpClient {
    fn drop(&mut self) {
        self.stop();
    }
}
